rfc/2019-12-06-dumpling.md

```diff
@@ -0,0 +1,264 @@
+- Last updated:  Dec 06, 2019
+- Discussion at: https://github.com/pingcap/community/issues/122, https://github.com/pingcap/community/pull/123
+
+## Abstract
+
+We propose introduce a library to replace Mydumper, code named **Dumpling**,
+optimized for TiDB Lightning and to be usable as a library/plugin inside DM and
+TiDB, as well as be an independent program.
+
+## Background
+
+### Optimize output format for machine reading
+
+[Mydumper] is a tool to dump MySQL databases into local filesystem as SQL dump.
+
+[TiDB Lightning] (a data importing tool to TiDB) and [DM] (a platform for
+migration data from MySQL to TiDB) currently relies on Mydumper to extract data
+from an upstream database.
+
+[TiDB Lightning]: https://pingcap.com/docs/dev/reference/tools/tidb-lightning/overview/
+[DM]: https://pingcap.com/docs/dev/reference/tools/data-migration/overview/
+[Mydumper]: https://github.com/pingcap/mydumper/
+
+Currently Lightning needs a [relatively complex tokenizer][t] to parse the SQL
+dump. This slows down the import speed. If we could make the data export tool to
+produce a very easy-to-parse binary output, the processing speed could be
+improved.
+
+[t]: https://github.com/pingcap/tidb-lightning/blob/master/lightning/mydump/parser.rl
+
+However, we do not want to heavily modify Mydumper because of the next reason.
+
+### License and maintenance
+
+Mydumper is licensed in GPLv3. This is incompatible with the license of all
+other PingCAP projects (Apache 2.0). This forces mydumper to only be usable as
+an external program in DM, instead of a library (neither static nor dynamic).
+
+Licensing aside, Mydumper is written in C and GLib, and does not expose itself
+as a library (almost the entire program is included in a single file). This
+further hinders us from tightly integrating it into DM and TiDB, which would
+have expected Go modules.
+
+Finally, the [official Mydumper repository] is sparsely updated, and we prefer
+not to create too much divergence in our fork. These prevent us from investing
+into the Mydumper project, and drive us to create a brand new project.
+
+[official Mydumper repository]: https://github.com/maxbube/mydumper
+
+### New features
+
+As a new project designed for the TiDB ecosystem, we can include further
+features we've desired for long, e.g. writing directly to cloud storage.
+
+## Proposal
+
+We propose introducing a new project **Dumpling**, which is
+
+* Primarily a library written in Go.
+* Support writing SQL dumps from MySQL-compatible database,
+    with speed and concurrency matching Mydumper.
+* Support every feature of Mydumper needed by DM and Lightning, but no more.
+* Support writing in SQL, CSV, and a custom binary format for quick consumption.
+* Support writing to cloud storage besides the local filesystem.
+
+## Rationale
+
+### Name
+
+The initial motivation of this tool is to supplement Lightning.
+We call the new tool "Dumpling" as a portmanteau of "dump" + "Lightning".
```

Comment:
TiDumpling might help with both searchability and understanding. There are already open source projects named dumpling.

Reply:
Naming is hard 🙃. We could also specify the official name as **TiDB Dumpling** (like TiDB Lightning and TiDB Binlog).